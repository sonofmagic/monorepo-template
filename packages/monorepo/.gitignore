assets
fixtures/assets
test/fixtures/demo
templates