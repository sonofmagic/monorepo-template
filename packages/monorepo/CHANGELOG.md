# @icebreakers/monorepo

## 0.6.13

### Patch Changes

- [`0196b4a`](https://github.com/sonofmagic/monorepo-template/commit/0196b4a79fd51577e6aadfeb0239c00a49b9fbb3) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade

## 0.6.3

### Patch Changes

- [`1b1f95e`](https://github.com/sonofmagic/monorepo-template/commit/1b1f95e6f67666b95c3844187a030a942e514ae1) Thanks [@sonofmagic](https://github.com/sonofmagic)! - feat: 默认不覆盖 workspace: 注册的包，cnpm sync 的时候，默认跳过 root 的包，只包括那些 workspace

## 0.5.0

### Minor Changes

- [`5b9a48f`](https://github.com/sonofmagic/monorepo-template/commit/5b9a48f0037aed5fc29520cbc91b5be59e647ee3) Thanks [@sonofmagic](https://github.com/sonofmagic)! - 内置整个 cli 功能，去除 scripts/monorepo 相关功能

## 0.4.6

### Patch Changes

- [`d85811a`](https://github.com/sonofmagic/monorepo-template/commit/d85811a60ef85df3beb2a2e0c9bdb332d1656e6a) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: github action ci/pr error
