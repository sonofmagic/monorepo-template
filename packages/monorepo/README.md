# @icebreakers/monorepo

[icebreaker](https://github.com/sonofmagic) 的 `monorepo` 升级同步工具

## 使用方式

```sh
# 安装
pnpm add -D @icebreakers/monorepo@latest
# 升级
npx monorepo up
# 帮助文档
npx monorepo -h
```

## 文档地址

https://monorepo.icebreaker.top/

## 需求环境

Nodejs >= `v20.11.0`
