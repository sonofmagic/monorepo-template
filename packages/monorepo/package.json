{
  "name": "@icebreakers/monorepo",
  "type": "module",
  "version": "0.6.13",
  "description": "icebreaker's monorepo config generator",
  "author": "ice breaker <1324318532@qq.com>",
  "license": "MIT",
  "homepage": "https://monorepo.icebreaker.top",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sonofmagic/monorepo-template.git",
    "directory": "packages/monorepo"
  },
  "bugs": {
    "url": "https://github.com/sonofmagic/monorepo-template/issues"
  },
  "keywords": [
    "monorepo",
    "icebreaker"
  ],
  "sideEffects": false,
  "exports": {
    ".": "./src/index.ts"
  },
  "bin": "dev/bin.js",
  "files": [
    "assets",
    "bin",
    "dist",
    "templates"
  ],
  "scripts": {
    "dev": "tsup --watch --sourcemap",
    "build": "tsup",
    "test": "pnpm prepublish && vitest run",
    "test:dev": "vitest",
    "release": "pnpm publish",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix",
    "prepublish": "tsx scripts/prepublish.ts"
  },
  "publishConfig": {
    "exports": {
      ".": {
        "types": "./dist/index.d.ts",
        "import": "./dist/index.js",
        "require": "./dist/index.cjs"
      }
    },
    "main": "./dist/index.cjs",
    "module": "./dist/index.js",
    "types": "./dist/index.d.ts",
    "bin": "bin/copy.js"
  },
  "dependencies": {
    "@inquirer/checkbox": "^4.0.1",
    "@inquirer/confirm": "^5.0.1",
    "@pnpm/logger": "^5.2.0",
    "@pnpm/types": "^12.2.0",
    "@pnpm/workspace.find-packages": "^4.0.11",
    "commander": "^12.1.0",
    "comment-json": "^4.2.5",
    "consola": "^3.2.3",
    "execa": "^9.4.1",
    "fs-extra": "^11.2.0",
    "git-url-parse": "^15.0.0",
    "klaw": "^4.1.0",
    "p-queue": "^8.0.1",
    "pathe": "^1.1.2",
    "picocolors": "^1.1.1",
    "simple-git": "^3.27.0"
  }
}
