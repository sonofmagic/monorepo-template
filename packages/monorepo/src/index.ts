export * from './monorepo'
export * from './utils'
