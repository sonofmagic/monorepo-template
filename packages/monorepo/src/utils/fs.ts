/**
 * 判断是否为可忽略的文件系统错误。
 * - ENOENT: 文件已被删除
 * - EBUSY: Windows 中资源被系统占用
 */
export function isIgnorableFsError(error: unknown): error is NodeJS.ErrnoException {
  if (!error) {
    return false
  }
  const code = (error as NodeJS.ErrnoException).code
  return code === 'ENOENT' || code === 'EBUSY' || code === 'EEXIST'
}
