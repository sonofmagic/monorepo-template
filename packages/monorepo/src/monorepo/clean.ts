import fs from 'fs-extra'
import path from 'pathe'
import set from 'set-value'

const cleanDirs = [
  'packages/monorepo',
  'packages/tsup-template',
  'packages/vue-lib-template',
  // 'packages/foo',
  // 'apps/cli',
  // 'apps/website',
  'apps',
]

export async function cleanProjects(cwd: string) {
  for (const dir of cleanDirs.map((x) => {
    return path.resolve(cwd, x)
  })) {
    await fs.remove(dir)
  }
  const name = 'package.json'
  const pkgJson = await fs.readJson(name)
  // fix https://github.com/sonofmagic/monorepo-template/issues/76
  set(pkgJson, 'devDependencies.@icebreakers/monorepo', 'latest', { preservePaths: false })
  await fs.outputJson(name, pkgJson, {
    spaces: 2,
  })
}
