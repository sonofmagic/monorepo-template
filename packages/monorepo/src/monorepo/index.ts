import type { CreateNewProjectOptions } from './create'
import type { GetWorkspacePackagesOptions } from './workspace'
import { cleanProjects } from './clean'
import { createNewProject } from './create'
import { GitClient } from './git'
import { init } from './init'
import { setVscodeBinaryMirror } from './mirror'
import { syncNpmMirror } from './sync'
import { upgradeMonorepo } from './upgrade'
import { getWorkspaceData, getWorkspacePackages } from './workspace'

export type {
  CreateNewProjectOptions,
  GetWorkspacePackagesOptions,
}

export {
  cleanProjects,
  createNewProject,
  getWorkspaceData,
  getWorkspacePackages,
  GitClient,
  init,
  setVscodeBinaryMirror,
  syncNpmMirror,
  upgradeMonorepo,
}
