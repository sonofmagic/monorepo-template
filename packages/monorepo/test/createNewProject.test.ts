import { isCI } from 'ci-info'
import { fdir as Fdir } from 'fdir'
import fs from 'fs-extra'
import path from 'pathe'
import { createNewProject, upgradeMonorepo } from '@/monorepo'

async function scanFiles(root: string) {
  const api = new Fdir().withRelativePaths()

  return (await api.crawl(root).withPromise()).sort((a, b) => a.localeCompare(b))
}

describe.skipIf(isCI)('createNewProject', () => {
  beforeAll(async () => {
    await import('../scripts/prepublish')
    await fs.remove(path.resolve(__dirname, './fixtures/demo'))
  })

  it('assets', async () => {
    const files = await scanFiles(path.resolve(__dirname, '../assets'))
    expect(files).toMatchSnapshot()
  })

  it('upgradeMonorepo', async () => {
    await upgradeMonorepo({
      outDir: path.resolve(__dirname, './fixtures/demo/upgrade'),
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/upgrade'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo unbuild case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/case-unbuild',
      renameJson: true,
      type: 'unbuild',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/case-unbuild'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo tsup case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/case-tsup',
      renameJson: true,
      type: 'tsup',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/case-tsup'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo vue-ui case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/case-vue-ui',
      renameJson: true,
      type: 'vue-lib',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/case-vue-ui'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo default case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/case-default',
      renameJson: true,
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/case-default'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo cli case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/cli',
      renameJson: true,
      type: 'cli',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/cli'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo client case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/client',
      renameJson: true,
      type: 'vue-hono',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/client'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo server case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/server',
      renameJson: true,
      type: 'hono-server',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/server'))
    expect(files).toMatchSnapshot()
  })

  it('createNewProject demo website case', async () => {
    await createNewProject({
      cwd: path.resolve(__dirname, './fixtures'),
      name: 'demo/website',
      renameJson: true,
      type: 'vitepress',
    })
    const files = await scanFiles(path.resolve(__dirname, './fixtures/demo/website'))
    expect(files).toMatchSnapshot()
  })
})
