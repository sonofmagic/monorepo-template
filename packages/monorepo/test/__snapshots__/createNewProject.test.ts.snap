// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createNewProject > assets 1`] = `
[
  ".changeset/config.json",
  ".changeset/few-islands-wish.md",
  ".changeset/fine-ravens-admire.md",
  ".changeset/fuzzy-dogs-study.md",
  ".changeset/real-actors-heal.md",
  ".changeset/spicy-boxes-worry.md",
  ".changeset/spicy-things-chew.md",
  ".changeset/tame-hairs-wait.md",
  ".dockerignore",
  ".editorconfig",
  ".gitattributes",
  ".github/FUNDING.yml",
  ".github/ISSUE_TEMPLATE/bug_report.yml",
  ".github/ISSUE_TEMPLATE/config.yml",
  ".github/workflows/ci.yml",
  ".github/workflows/release.yml",
  ".gitignore",
  ".husky/commit-msg",
  ".husky/pre-commit",
  ".npmrc",
  ".vscode/extensions.json",
  ".vscode/settings.json",
  "CODE_OF_CONDUCT.md",
  "commitlint.config.ts",
  "CONTRIBUTING.md",
  "Dockerfile",
  "eslint.config.js",
  "gitignore",
  "LICENSE",
  "lint-staged.config.js",
  "netlify.toml",
  "package.json",
  "pnpm-workspace.yaml",
  "renovate.json",
  "SECURITY.md",
  "stylelint.config.js",
  "tsconfig.json",
  "turbo.json",
  "vitest.config.ts",
  "vitest.workspace.ts",
]
`;

exports[`createNewProject > createNewProject demo cli case 1`] = `
[
  ".gitignore",
  "bin/index.js",
  "dev/index.ts",
  "package.mock.json",
  "src/cli.ts",
  "src/index.ts",
  "tsconfig.json",
  "tsup.config.ts",
]
`;

exports[`createNewProject > createNewProject demo client case 1`] = `
[
  ".gitignore",
  "eslint.config.js",
  "index.html",
  "package.mock.json",
  "public/vite.svg",
  "src/App.vue",
  "src/main.ts",
  "src/pages/index.vue",
  "src/router/index.ts",
  "src/stores/i18n.ts",
  "src/stores/index.ts",
  "src/style.css",
  "src/trpc/index.ts",
  "src/vite-env.d.ts",
  "tsconfig.app.json",
  "tsconfig.json",
  "tsconfig.node.json",
  "tsconfig.worker.json",
  "vite.config.ts",
  "worker/app.ts",
  "worker/config.ts",
  "worker/fetch-entry.ts",
  "worker/node-entry.ts",
  "worker/trpc-router.ts",
  "wrangler.jsonc",
]
`;

exports[`createNewProject > createNewProject demo default case 1`] = `
[
  "build.config.ts",
  "package.mock.json",
  "src/index.ts",
  "src/utils.ts",
  "test/index.test.ts",
  "tsconfig.json",
  "vitest.config.ts",
]
`;

exports[`createNewProject > createNewProject demo server case 1`] = `
[
  ".gitignore",
  "package.mock.json",
  "src/app.ts",
  "src/config.ts",
  "src/fetch-entry.ts",
  "src/node-entry.ts",
  "test/index.test.ts",
  "tsconfig.json",
  "tsup.config.ts",
  "vitest.config.ts",
  "wrangler.jsonc",
]
`;

exports[`createNewProject > createNewProject demo tsup case 1`] = `
[
  "build.config.ts",
  "package.mock.json",
  "src/index.ts",
  "test/index.test.ts",
  "tsconfig.json",
  "tsup.config.ts",
  "vitest.config.ts",
]
`;

exports[`createNewProject > createNewProject demo unbuild case 1`] = `
[
  "build.config.ts",
  "package.mock.json",
  "src/index.ts",
  "src/utils.ts",
  "test/index.test.ts",
  "tsconfig.json",
  "vitest.config.ts",
]
`;

exports[`createNewProject > createNewProject demo vue-ui case 1`] = `
[
  ".gitignore",
  "eslint.config.js",
  "index.html",
  "lib/HelloWorld.vue",
  "lib/index.ts",
  "package.mock.json",
  "src/App.vue",
  "src/main.ts",
  "src/pages/index.vue",
  "src/router/index.ts",
  "src/style.css",
  "src/vite-env.d.ts",
  "test/index.test.ts",
  "tsconfig.app.json",
  "tsconfig.json",
  "tsconfig.node.json",
  "tsconfig.test.json",
  "vite.config.ts",
  "vite.shared.config.ts",
  "vitest.config.ts",
]
`;

exports[`createNewProject > createNewProject demo website case 1`] = `
[
  ".gitignore",
  ".vitepress/config.ts",
  ".vitepress/theme/index.ts",
  ".vitepress/theme/tailwind.css",
  "index.md",
  "monorepo/index.md",
  "monorepo/manage.md",
  "monorepo/publish.md",
  "package.mock.json",
  "public/logo.jpg",
  "thinking.md",
  "tools/changeset.md",
  "tools/husky.md",
  "tools/lint-staged.md",
  "tools/pnpm.md",
  "tools/renovate.md",
  "tools/turborepo.md",
  "why/assets/npm-dt.svg",
  "why/assets/npm-ts.svg",
  "why/examples/0.npm-basic-package/index.js",
  "why/examples/0.npm-basic-package/package.json",
  "why/examples/1.npm-basic-package/index.cjs",
  "why/examples/1.npm-basic-package/index.d.ts",
  "why/examples/1.npm-basic-package/index.js",
  "why/examples/1.npm-basic-package/package.json",
  "why/examples/2.npm-basic-package/package-lock.json",
  "why/examples/2.npm-basic-package/package.json",
  "why/examples/2.npm-basic-package/README.md",
  "why/examples/2.npm-basic-package/rename-ext.js",
  "why/examples/2.npm-basic-package/src/index.ts",
  "why/examples/2.npm-basic-package/tsconfig.cjs.json",
  "why/examples/2.npm-basic-package/tsconfig.esm.json",
  "why/examples/2.npm-basic-package/tsconfig.json",
  "why/examples/3.npm-basic-package/build.config.ts",
  "why/examples/3.npm-basic-package/package-lock.json",
  "why/examples/3.npm-basic-package/package.json",
  "why/examples/3.npm-basic-package/README.md",
  "why/examples/3.npm-basic-package/src/index.ts",
  "why/examples/3.npm-basic-package/tsconfig.json",
  "why/examples/3.npm-basic-package/tsup.config.ts",
  "why/examples/usage-cjs/cjs.cjs",
  "why/examples/usage-cjs/esm.mjs",
  "why/examples/usage-cjs/index.js",
  "why/examples/usage-cjs/package-lock.json",
  "why/examples/usage-cjs/package.json",
  "why/examples/usage-esm/cjs.cjs",
  "why/examples/usage-esm/createRequire.js",
  "why/examples/usage-esm/esm.mjs",
  "why/examples/usage-esm/index.js",
  "why/examples/usage-esm/package-lock.json",
  "why/examples/usage-esm/package.json",
  "why/how-to-reuse-js-code.md",
  "why/index.md",
  "why/js-cjs-mjs.md",
  "why/js-keywords.md",
  "why/modern/bundlers.md",
  "why/modern/dts.md",
  "why/modern/esm-vs-cjs.md",
  "why/modern/index.md",
  "why/modern/package-entry-points.md",
  "why/modern/typescript.md",
  "why/publish-basic-npm-package.md",
  "why/what-is-npm-package.md",
]
`;

exports[`createNewProject > upgradeMonorepo 1`] = `
[
  ".changeset/config.json",
  ".changeset/few-islands-wish.md",
  ".changeset/fine-ravens-admire.md",
  ".changeset/fuzzy-dogs-study.md",
  ".changeset/real-actors-heal.md",
  ".changeset/spicy-boxes-worry.md",
  ".changeset/spicy-things-chew.md",
  ".changeset/tame-hairs-wait.md",
  ".dockerignore",
  ".editorconfig",
  ".gitattributes",
  ".github/FUNDING.yml",
  ".github/ISSUE_TEMPLATE/bug_report.yml",
  ".github/ISSUE_TEMPLATE/config.yml",
  ".github/workflows/ci.yml",
  ".github/workflows/release.yml",
  ".gitignore",
  ".husky/commit-msg",
  ".husky/pre-commit",
  ".npmrc",
  ".vscode/extensions.json",
  ".vscode/settings.json",
  "CODE_OF_CONDUCT.md",
  "commitlint.config.ts",
  "CONTRIBUTING.md",
  "Dockerfile",
  "eslint.config.js",
  "LICENSE",
  "lint-staged.config.js",
  "netlify.toml",
  "pnpm-workspace.yaml",
  "renovate.json",
  "SECURITY.md",
  "stylelint.config.js",
  "tsconfig.json",
  "turbo.json",
  "vitest.config.ts",
]
`;
